lib/TPP/Conversion/ConvertVectorToXsmm/ConvertVectorToXsmm.cpp

```diff
@@ -0,0 +1,220 @@
+  IntegerAttr unaryFlags = dyn_cast<IntegerAttr>(
+      xsmm::UnaryFlagsAttr::get(rewriter.getContext(), xsmm::UnaryFlags::NONE));
+  dispatchOperands.push_back(unaryFlags.getInt());
+
+  auto dispatchCall = xsmm::utils::buildXsmmCall(
+      rewriter, xsmm::utils::XsmmCallType::DISPATCH, loc, dtype,
+      dispatchOperands, IntegerType::get(rewriter.getContext(), 64),
+      SymbolRefAttr::get(transposeOp->getContext(), dispatchName), transposeOp,
+      nullptr);
+  SmallVector<xsmm::utils::XsmmOperand> operandRange{
+      dyn_cast<DataTypeAttr>(dtype).getInt(),
+      xsmm::utils::XsmmCall{xsmm::utils::XsmmCallType::DISPATCH, dispatchCall},
+      input->getOperand(0), output->getOperand(1)};
+  auto invokeCall = xsmm::utils::buildXsmmCall(
+      rewriter, xsmm::utils::XsmmCallType::INVOKE, loc, dtype, operandRange,
+      TypeRange(), SymbolRefAttr::get(transposeOp->getContext(), invokeName),
+      transposeOp, output);
+  return std::make_pair(&*dispatchCall, &*invokeCall);
+}
+
+static std::pair<Operation *, Operation *>
+convertTransposeOp(PatternRewriter &rewriter, Operation *transposeOp,
+                   Operation *input, Operation *output, Type outputType) {
+  LLVM_DEBUG(llvm::dbgs() << "convertTransposeOp\n");
+  VectorType outType = cast<VectorType>(outputType);
+  xsmm::UnaryKind opType;
+  xsmm::UnaryInfo unaryInfo;
+  MemRefType firstOperand;
+  MemRefType shapedType;
+
+  if (vnni::utils::isInVnniLayout(vnni::utils::VnniOperandRank::TRANSPOSE,
+                                  outType)) {
+    memref::ExpandShapeOp expandShapeOp =
+        dyn_cast<memref::ExpandShapeOp>(input->getOperand(0).getDefiningOp());
+    firstOperand = expandShapeOp.getSrcType();
+    shapedType = expandShapeOp.getSrcType();
+    opType = xsmm::UnaryKind::VNNI2;
+  } else {
+    firstOperand = dyn_cast<MemRefType>(input->getOperand(0).getType());
+    shapedType = dyn_cast<MemRefType>(output->getOperand(1).getType());
+    opType = xsmm::UnaryKind::TRANSPOSE;
+  }
+
+  unaryInfo = *xsmm::utils::getVectorUnaryInfo(
+      shapedType, firstOperand,
+      dyn_cast<MemRefType>(output->getOperand(1).getType()),
+      dyn_cast<VectorType>(input->getResult(0).getType()),
+      dyn_cast<VectorType>(output->getOperand(0).getType()),
+      xsmm::UnaryFlags::NONE);
+  if (vnni::utils::isInVnniLayout(vnni::utils::VnniOperandRank::TRANSPOSE,
+                                  outType)) {
+    // Adjust ldi and ldo  based on the VNNI factor.
+    auto vnniFactor = *vnni::utils::getVnniBlockingFactor(outType);
+    unaryInfo.ldo = unaryInfo.ldo * vnniFactor;
+    unaryInfo.ldi = unaryInfo.ldi / vnniFactor;
+    std::swap(unaryInfo.ldi, unaryInfo.ldo);
+  } else
+    std::swap(unaryInfo.m, unaryInfo.n);
+  return getTransposeOperation(rewriter, unaryInfo, outputType, opType, input,
+                               output, transposeOp);
+}
+
+static LogicalResult
+validateTransposeOpImpl(PatternRewriter &rewriter, Operation *transposeOp,
```

Comment:
Why `Impl` in the name? This is a local function that just validates something. `Impl` is a pattern in LLVM/MLIR (and general C++ projects) that imply it's the _hidden_ part of a public API that is not accessible to the users of that API.

Here, there's no API.